### PR TITLE
Allow limiting of domains for OIDC (Fixes #3709)

### LINKED ISSUE #3709: Allow restricting of domains for OIDC auth
Take the example of Implicit Auth with Google - with `register:true`, a valid Google user will automatically create a Sync Gateway user with the name like `accounts.google.com_01189998819991197253`. Obviously, this is viable for authentication as the user is definitely (in so far as we can trust Google) who they say they are, and by setting `register:true` we've opted in to creating a user record based on that fact. 

We haven't yet authorised the user to access anything other than the public channel (`!`), so there's not really a risk here. It does mean though that we have to add an extra step (and, likely, component) - it would be nice if we could do some of this inside Sync Gateway.

Imagine creating a application/service where we want to allow any user from one company to create an account, but no one else. If we can validate that the email address for the user we just authenticated ends with `@couchbase.com` and prevent any others from registering, we can treat the public channel as authenticated as it's restricted to `@couchbase.com` users.

### PR DESCRIPTION
**Pull request by JFlath**
Recognise that this probably needs a bit of work - opening this for (hopefully somewhat gentle ;)) feedback, and because I wanted a barebones implementation for myself anyway...

### CODE REVIEW

File: auth/oidc_test.go [hunk 1]
```diff
@@ -43,6 +43,36 @@ func TestOIDCUsername(t *testing.T) {
...
+	provider := OIDCProvider{
+		Issuer: "http://www.someprovider.com",
+		WhitelistDomains: []string{"couchbase.com", "membase.com"},
+	}
+
+	provider.InitUserPrefix()
+
+	err := ValidateAllowedDomain(&provider, "jflath@couchbase.com")
+	assert.Equals(t, err, nil)
+	err = ValidateAllowedDomain(&provider, "jflath@gmail.com")
```
Comment (JFlath) [inline auth/oidc_test.go]:
```suggestion
	err = ValidateAllowedDomain(&provider, "jflath@example.com")
```

File: auth/oidc_test.go [hunk 2]
```diff
@@ -43,6 +43,36 @@ func TestOIDCUsername(t *testing.T) {
...
+	provider := OIDCProvider{
+		Issuer: "http://www.someprovider.com",
+		BlacklistDomains: []string{"gmail.com", "hotmail.com"},
+	}
+
+	provider.InitUserPrefix()
+
+	err := ValidateAllowedDomain(&provider, "jflath@couchbase.com")
+	assert.Equals(t, err, nil)
+	err = ValidateAllowedDomain(&provider, "jflath@gmail.com")
```
Comment (JFlath) [inline auth/oidc_test.go]:
```suggestion
	err = ValidateAllowedDomain(&provider, "jflath@example.com")
```

File: auth/oidc_test.go [hunk 3]
```diff
@@ -43,6 +43,36 @@ func TestOIDCUsername(t *testing.T) {
...
+	assert.Equals(t, err, nil)
+	err = ValidateAllowedDomain(&provider, "jflath@example.com")
+	assert.NotEquals(t, err, nil)
+}
+
+func TestOIDCBlacklist(t *testing.T) {
+
+	provider := OIDCProvider{
+		Issuer: "http://www.someprovider.com",
+		BlacklistDomains: []string{"gmail.com", "hotmail.com"},
```
Comment (JFlath) [inline auth/oidc_test.go]:
```suggestion
		BlacklistDomains: []string{"gmail.com", "example.com"},
```

**Comment by adamcfraser** [comment]
@JFlath I'm not yet convinced that the underlying issue strictly requires a fix.  If this is still critical from your perspective, can you schedule a call to discuss?

**Pull request closed without merging**